### DISCUSSION TITLE
Self regulation tuning question

### DISCUSSION THREAD

**Discussion by Edwintenhaaf** [jmcollin78/versatile_thermostat#380] [Q&A] [answered]
Hi All,

I am trying to create the perfect setting for self regulation but it does not work out verry well.
Using: thermostat_over_climate
Have read the manual about [this](https://github.com/jmcollin78/versatile_thermostat?tab=readme-ov-file#self-regulation-in-expert-mode) but I do not understand what settings I need to adjust.
This currently how it looks in the graph.
<img width="416" alt="image" src="https://github.com/jmcollin78/versatile_thermostat/assets/51457250/15c7d903-cf5e-4f59-bfaf-3125d352f687">
If the temperature drops the thermostat responds to slow and sometimes it gets more the 1 degree colder in the living room.
How can I change this behavior ?

Currently experimenting with expert mode using the following settings ?
```
versatile_thermostat:
    auto_regulation_expert:
      kp: 0.4
      ki: 0.08
      k_ext: 1.0
      offset_max: 5
      stabilization_threshold: 0.1
      accumulated_error_threshold: 55
```
      
 In the notes of the 'slow regulation' there is this part:
`ki: 0.8 / 288.0  # 80% of the current internal regulation offset are caused by the average offset of the past 24 hours`
Is this average only calculated when the AC is turned on? I turn off the AC during the night so the average could not be correct.

Thanks,

Edwin

**Comment by jmcollin78** [comment]
Hello @Edwintenhaaf , Can you post the graphs with this installed ? https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#even-better-with-plotly-to-tune-your-thermostat

And I need also the VTherm attributes (Development Tools / States copy/paste attributes and format with the <> button).

> Is this average only calculated when the AC is turned on? I turn off the AC during the night so the average could not be correct.

No temperature average are always calculated. We only need to have the temperature sensor working for this.

What is your device ? Is it a air/air heat pump ?

**Reply by Edwintenhaaf** [reply]
Hi @jmcollin78,

The device is a air/air heatpump:
Toshiba RAS-3M18G3AVG ( outside unit)
2x RAS-B13G3KVSG (living room & first floor)
1x RAS-B 10 J2FVG (attic)

After my post I have experiment a litle bit more and it looks a bit better this morning.
The house is not optimal isolated so the temperature drops quickly in winter and rises fast in summer.
Current outside temperatures between 5 - 10 degrees celsius.
I have also checked if 'defrosts' of the ac could be the issue. But that is not the case.

current settings:
```
versatile_thermostat:
    auto_regulation_expert:
      kp: 0.9
      ki: 0.15
      k_ext: 0.1
      offset_max: 5
      stabilization_threshold: 0.1
      accumulated_error_threshold: 75
```

Plotly
<img width="418" alt="image" src="https://github.com/jmcollin78/versatile_thermostat/assets/51457250/759391d7-4d62-47e7-8fde-c35f17883e8f">


```
hvac_modes:
  - "off"
  - auto
  - cool
  - heat
  - dry
  - fan_only
min_temp: 15
max_temp: 30
target_temp_step: 1
fan_modes:
  - Auto
  - Quiet
  - Low
  - Medium Low
  - Medium
  - Medium High
  - High
preset_modes:
  - none
  - frost
  - eco
  - comfort
  - boost
swing_modes:
  - "Off"
  - Swing Vertical
  - Swing Horizontal
  - Swing Vertical And Horizontal
  - Fixed 1
  - Fixed 2
  - Fixed 3
  - Fixed 4
  - Fixed 5
current_temperature: 19.8
temperature: 20
fan_mode: Medium
preset_mode: none
swing_mode: Fixed 5
is_on: true
hvac_action: null
hvac_mode: heat
type: null
is_controlled_by_central_mode: false
last_central_mode: null
frost_temp: 15
eco_temp: 19
boost_temp: 22
comfort_temp: 21
frost_away_temp: 30
eco_away_temp: 18
boost_away_temp: 18
comfort_away_temp: 18
power_temp: 13
target_temperature_step: 1
ext_current_temperature: 9.7
ac_mode: true
current_power: null
current_power_max: null
saved_preset_mode: none
saved_target_temp: 21
saved_hvac_mode: null
motion_sensor_entity_id: null
motion_state: null
power_sensor_entity_id: null
max_power_sensor_entity_id: null
overpowering_state: null
presence_sensor_entity_id: null
presence_state: null
window_state: "off"
window_auto_state: "off"
window_bypass_state: false
window_sensor_entity_id: null
window_delay_sec: 300
window_auto_enabled: false
window_auto_open_threshold: null
window_auto_close_threshold: null
window_auto_max_duration: null
window_action: window_turn_off
security_delay_min: 180
security_min_on_percent: 0.5
security_default_on_percent: 0.8
last_temperature_datetime: "2024-02-14T07:49:54.675392+01:00"
last_ext_temperature_datetime: "2024-02-14T07:50:10.853188+01:00"
security_state: false
minimal_activation_delay_sec: 10
device_power: 0
mean_cycle_power: null
total_energy: 0
last_update_datetime: "2024-02-14T07:50:10.859147+01:00"
timezone: Europe/Amsterdam
temperature_unit: °C
is_device_active: true
ema_temp: 19.69
is_used_by_central_boiler: false
is_over_climate: true
start_hvac_action_date: null
underlying_climate_0: climate.woonkamer_ac
underlying_climate_1: null
underlying_climate_2: null
underlying_climate_3: null
is_regulated: true
regulated_target_temperature: 21
auto_regulation_mode: auto_regulation_expert
regulation_accumulated_error: 26.444501953125013
auto_fan_mode: auto_fan_none
current_auto_fan_mode: auto_fan_none
auto_activated_fan_mode: null
auto_deactivated_fan_mode: null
auto_regulation_use_device_temp: true
friendly_name: Woonkamer Airconditioning
supported_features: 441
```

**Reply by jmcollin78** [reply]
ok thanks. So:
1. you use the new "auto_regulation_use_device_temp: true", to compensate the internal temp sensor. This is fine,
2. the EMA follows the room temp and seems good,
3. regulated temps is much over the target, so it works as expected,

So all seems to works fine but the room temp is often under the target. I think this is because the AC stops heating to early and I don't know why now.
Can you please, add the AC internal temp and AC current temp in the Plotly grah (and remove EMA, and slope which are fine) ?

**Comment by Edwintenhaaf** [comment]
If the target temp is reached the regulated temp almost immediately drops
The modified graph
<img width="492" alt="image" src="https://github.com/jmcollin78/versatile_thermostat/assets/51457250/a5db3a07-c98d-46a3-bcd0-c2ddbdfc1269">

**Reply by jmcollin78** [reply]
The first part (00:00-> 07:00) is very, very strange. Did you turn the AC off ?
On the second part, what I see in not too bad. There is an oscillation around the target. This is not totally unusual:
![Capture d’écran 2024-02-14 à 14 17 57](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/588bc2a8-320e-4d84-ba84-608638cb0fa0)

Your problem seems that there is no inertia at all. The temperature fails immediatly when heating stops, and increase immediatly after heating. It is the first time I see that. Is that a very small room and very badly isolated ?

Where is the regulated temp ?

**Reply by Edwintenhaaf** [reply]
Yes I turn off the ac during the night. 
Tried leaving it on but from a cost perspective a bad idea.

The living room is 7.30mx3.70mx2.55m in size 
The attached small kitchen is 3.10mx2.40mx2.55m

It is isolated but depending on outside temperature and wind direction it gets cold quickly.
These are [typical dutch houses](https://www.dutchnews.nl/2022/08/terraced-homes-dominate-dutch-housing-sector/) with large windows. Second floor and attic are the warmest parts of the house. The floors are made of wood and not isolated.

Thanks for you're help I will keep trying to get it working!

The graph looks better today since changing the parameters
<img width="422" alt="image" src="https://github.com/jmcollin78/versatile_thermostat/assets/51457250/ab93cb33-0b7f-4a58-b1cc-71bdf5c49886">

```
versatile_thermostat:
    auto_regulation_expert:
      kp: 0.9
      ki: 0.15
      k_ext: 1.5
      offset_max: 5
      stabilization_threshold: 0.1
      accumulated_error_threshold: 75
```

The regulated temp is hidden behind the ac target temp
`auto_regulation_use_device_temp: true`

**Reply by jmcollin78** [reply]
> The regulated temp is hidden behind the ac target temp
auto_regulation_use_device_temp: true

This should not ! The target temp send to the target should be the regulated temp + the offset (room temp - device temp).

So I guess, your device temp is not available which certainly explains that heating is so short.
You should have a `current_temperature` attribute in your AC device. You can check this in the Development Tools / State.
Can you post me those attributes ?

**Reply by jmcollin78** [reply]
And because there is no logs, I think nobody is able to see that:

```

@property
    def underlying_current_temperature(self) -> float | None:
        """Get the underlying current_temperature if it exists
        and if initialized"""
        if not self.is_initialized:
            return None

        if not hasattr(self._underlying_climate, "current_temperature"):
            return None

        return self._underlying_climate.current_temperature
```

**Reply by Edwintenhaaf** [reply]
Ahhh because of this [issue](https://github.com/jmcollin78/versatile_thermostat/discussions/375) I created an automation as a workaround to set the target temp on the ac
This does not add the offset you mentioned.
My mistake, sorry for that.

**Reply by jmcollin78** [reply]
You have implemented the regulation algo in an automation ? With the error_accumulation and all the coefs ?
You should have tell me that before ... I'm afraid I'm searching something that is not conform to the algo.

**Reply by jmcollin78** [reply]
To fix the bug of #375 if you want you can just change line 279 of thermostat_climate.py with these lines:
```
            await under.set_temperature(
                round_to_nearest(target_temp, self._auto_regulation_dtemp),
                self._attr_max_temp,
                self._attr_min_temp,
            )
```

This should fixe the bug (not tested on my side). Please respect indentation because Python is strict on this point.

Then please, remove your workaround, set the regulation algo to "Light", keep the `auto_regulation_use_device_temp` to true.
This should behave normaly and you should have a great regulation.

**Reply by Edwintenhaaf** [reply]
sorry but I did tell you that but it was in another discussion.
<img width="959" alt="image" src="https://github.com/jmcollin78/versatile_thermostat/assets/51457250/3ce39bdd-8c84-401b-b46a-1be5d9766efa">

**Reply by Edwintenhaaf** [reply]
Hi @jmcollin78 ,

changed the code but I might made some error because it still sends not rounded temps to the ac
Also did a HA restart after change.

![image](https://github.com/jmcollin78/versatile_thermostat/assets/51457250/f236f385-444f-47d9-bce7-916ae4ab695c)
```
Logger: homeassistant.core
Source: core.py:2295
First occurred: 7:10:46 AM (2 occurrences)
Last logged: 7:28:54 AM

Error executing service: <ServiceCall climate.set_temperature (c:01HPNP755VJXZT7CSC0K4GD58N): entity_id=['climate.woonkamer_ac'], temperature=21.97, target_temp_high=30.0, target_temp_low=15.0>
Error executing service: <ServiceCall climate.set_temperature (c:01HPNQ8BG554ZPZ2HTNCC56929): entity_id=['climate.woonkamer_ac'], temperature=20.52, target_temp_high=30.0, target_temp_low=15.0>
```

**Reply by jmcollin78** [reply]
ok. I will have to test this on my env (which is currently broken). Sorry, I though it could work with that.

**Reply by Edwintenhaaf** [reply]
no worries I tried to workaround this by doing the math in my automation and then send it to the ac for now.
If I can help testing new code please let me know.

   {## versatile thermostat ##}
    {% set ac_temp_reg = state_attr('climate.woonkamer','regulated_target_temperature')%} 
    {## ac measured temp ##}
    {% set ac_temp = state_attr('climate.woonkamer_ac','current_temperature') %}   
    {## room temp ##}
    {% set room_temp = states('sensor.toon_room_temp_for_ac_woonkamer') | float %} 
    {## The target temp send to the target should be the regulated temp + the offset (room temp - device temp ##}
    {% set offset = ac_temp - room_temp %}{% if offset < 0 %}{%set offset = 0%}{%endif%}
    {% set send_to_ac = ac_temp_reg + offset %}
    {{ send_to_ac | round }}

**Reply by Edwintenhaaf** [reply]
So far so good
![image](https://github.com/jmcollin78/versatile_thermostat/assets/51457250/c5bd9d4a-f729-470d-848e-3a470038c072)
Could not use the light parameters because the offset_max was 1.5 (only rounded works)
So made a copy and using expert mode
```
versatile_thermostat:
    auto_regulation_expert:
      kp: 0.2
      ki: 0.05
      k_ext: 0.05
      offset_max: 3
      stabilization_threshold: 0.1
      accumulated_error_threshold: 10
```

**Reply by jmcollin78** [reply, marked as answer]
Ok that is not so bad. The oscillation may be reduced (it depend of your device) by having a lower regulation parameters.
the red curve is too much above the target and is induce the oscillation. Maybe you can try to lower kp and/or the offset max.
Youhave a very dynamic system, so it could work.
